lib/spack/spack/cmd/checksum.py

```diff
@@ -30,93 +32,133 @@
 
+    sorted_versions = sorted(url_dict.keys(), reverse=True)
+
+    # Find length of longest string in the list for padding
+    max_len = max(len(str(v)) for v in sorted_versions)
+    num_ver = len(sorted_versions)
+
+    tty.msg("Found {0} version{1} of {2}:".format(
+            num_ver, '' if num_ver == 1 else '', name),
+            "",
+            *spack.cmd.elide_list(
+                ["{0:{1}}  {2}".format(v, max_len, url_dict[v])
+                 for v in sorted_versions]))
+    print()
+
+    archives_to_fetch = tty.get_number(
+        "How many would you like to checksum?", default=1, abort='q')
+
+    if not archives_to_fetch:
+        tty.die("Aborted.")
+
+    versions = sorted_versions[:archives_to_fetch]
+    urls = [url_dict[v] for v in versions]
+
     tty.msg("Downloading...")
-    hashes = []
+    version_hashes = []
     i = 0
     for url, version in zip(urls, versions):
         try:
             with Stage(url, keep=keep_stage) as stage:
+                # Fetch the archive
                 stage.fetch()
                 if i == 0 and first_stage_function:
+                    # Only run first_stage_function the first time,
+                    # no need to run it every time
                     first_stage_function(stage, url)
 
-                hashes.append((version, spack.util.crypto.checksum(
+                # Checksum the archive and add it to the list
+                version_hashes.append((version, spack.util.crypto.checksum(
                     hashlib.md5, stage.archive_file)))
                 i += 1
-        except FailedDownloadError as e:
-            tty.msg("Failed to fetch %s" % url)
+        except FailedDownloadError:
+            tty.msg("Failed to fetch {0}".format(url))
         except Exception as e:
-            tty.msg('Something failed on %s, skipping.\n    (%s)' % (url, e))
+            tty.msg("Something failed on {0}, skipping.".format(url),
+                    "  ({0})".format(e))
 
-    return hashes
+    if not version_hashes:
+        tty.die("Could not fetch any versions for {0}".format(name))
+
+    # Find length of longest string in the list for padding
+    max_len = max(len(str(v)) for v, h in version_hashes)
+
+    # Generate the version directives to put in a package.py
+    version_lines = "\n".join([
+        "    version('{0}', {1}'{2}')".format(
+            v, ' ' * (max_len - len(str(v))), h) for v, h in version_hashes
+    ])
+
+    num_hash = len(version_hashes)
+    tty.msg("Checksummed {0} version{1} of {2}:".format(
+        num_hash, '' if num_hash == 1 else 's', name))
+
+    return version_lines
 
 
 def checksum(parser, args):
-    # get the package we're going to generate checksums for
+    # Make sure the user provided a package and not a URL
+    if not valid_fully_qualified_module_name(args.package):
+        tty.die("`spack checksum` accepts package names, not URLs. "
+                "Use `spack md5 <url>` instead.")
```

Comment:
Does it make sense to rename `spack checksum`?  We will not always use `md5`, and it would be nice if `spack checksum` gave you whatever the current recommended checksum of a file/url was.  Should `spack checksum` be called `spack find-new-versions` or `spack-spider-versions` or something?